server/user.js

```diff
@@ -0,0 +1,182 @@
+/**
+ * @fileoverview user function handler
+ */
+var fs = require('fs');
+var assert = require('assert');
+
+var log = require('./log.js');
+var utils = require('./utils.js');
+
+/** @type {user} */
+module.exports = user;
+
+/**
+ * @constructor
+ */
+function user() {}
+
+/** @enum {string} */
+user.QueryType = {
+  SIGNUP: 'sign-up',
+  SIGNIN: 'sign-in'
+};
+
+/**
+ * Expire time of cookie.
+ * @const {number}
+ */
+user.cookieExpireTime = 24 * 60 * 60 * 1000;
+
+/**
+ * @typedef {{
+ *   email: (string|undefined),
+ *   username: string,
+ *   password: string,
+ *   confirmed: (boolean|undefined)
+ * }}
+ */
+user.Info;
+
+/**
+ * @typedef {{
+ *   error: string
+ * }}
+ */
+user.Error;
+
+/**
+ * Checks the user information for signing in.
+ * @param {!mongodb.Db} db Session database.
+ * @param {string} userPath Directory of user information file.
+ * @param {!user.Info} userInfo User Information
+ * @param {function(!Object)} callback Callback function.
+ */
+user.signUp = function(db, userPath, userInfo, callback) {
+  var checkDuplicate = {
+    duplicated: false,
+    elements: []
+  };
+  var cursor = db.collection('userInfo').find({
+    $or: [{username: userInfo.username}, {email: userInfo.email}]});
+  var data = [];
+  cursor.each(function(err, doc) {
+    assert.equal(err, null, 'error occurs');
+    if (doc != null) {
+      data.push(doc);
+    } else {
+      var result = authenticateCallback();
+      callback(result);
```

Comment:
if (err != null) {
   ...
}
if (count != 1) {
  ...
}


On Sun, Mar 27, 2016 at 1:13 AM, Bowen Yu <notifications@github.com> wrote:

> In server/user.js
> <https://github.com/ViDA-NYU/genotet/pull/145#discussion_r57523543>:
>
> > + */
> > +user.signUp = function(db, userPath, userInfo, callback) {
> > +  var checkDuplicate = {
> > +    duplicated: false,
> > +    elements: []
> > +  };
> > +  var cursor = db.collection('userInfo').find({
> > +    $or: [{username: userInfo.username}, {email: userInfo.email}]});
> > +  var data = [];
> > +  cursor.each(function(err, doc) {
> > +    assert.equal(err, null, 'error occurs');
> > +    if (doc != null) {
> > +      data.push(doc);
> > +    } else {
> > +      var result = authenticateCallback();
> > +      callback(result);
>
> I can't understand your comment "properties of cursor itslef are not
> accessible", as I never mentioned anything saying that we should access the
> cursor's properties.
>
> A few options are:
> 1. Comment that this cursor always contains exactly one element, so that
> the callback function will be called exactly once.
> 2. (*Probably a better way*) Create a wrapper function under db
> namespace, that performs a query and returns exactly 1 result.
>
> db.getOne = function(collection, query, callback) {
>   var cursor = collection.find(query);
>   // optional
>   cursor.count(function(err, count) {
>     if (err != null) {
>      ...
>     }
>     if (count != 1) {
>      ...
>     }
>   });
>   // count = 1
>   cursor.each(function(err, data) {
>     callback(data);
>   });
> };
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/ViDA-NYU/genotet/pull/145/files/d8d51e9309e87f211b37917377713fc1703002cc#r57523543>
>